src/scope.rs

```diff
@@ -4,13 +4,59 @@ use crate::{named::ProgExt, ProgNode};
 
 /// Tracker of variable bindings.
 ///
-/// Internally there is a stack of scopes.
-/// A new scope is pushed for each (nested) block expression.
+/// Each Simfony expression is compiled based on an _environment_ (Simplicity) value.
+/// The environment starts empty and is extended with each let statement.
+/// Variable names are resolved by extracting the corresponding value from the environment.
 ///
-/// Bindings from higher scopes (in the stack) overwrite bindings from lower scopes.
+/// Each (nested) block expression introduces a new scope.
+/// Bindings from inner scopes overwrite bindings from outer scopes.
+/// The bindings of a scope live for as long as the scope.
 #[derive(Debug, Clone)]
 pub struct GlobalScope {
-    variables: Vec<Vec<Pattern>>,
+    /// A stack of patterns which corresponds to a pattern tree.
+    /// New patterns are pushed onto the end of the stack.
+    ///
+    /// The empty stack `[]` corresponds to the tree `_`
+    /// with the ignore pattern at the root.
+    ///
+    /// The stack `[p1, p2, p3]` corresponds to a nested product of patterns:
+    ///
+    /// ```text
+    ///    .
+    ///   / \
+    /// p3   .
+    ///     / \
+    ///   p2   .
+    ///       / \
+    ///     p1   _
+    /// ```
+    ///
+    /// And so on.
+    ///
+    /// Later patterns occur higher in the tree.
+    /// Earlier patterns occur lower in the tree.
+    ///
+    /// The pattern tree is isomorphic to the current environment.
+    /// The isomorphism maps variable names from the pattern tree to values from the environment.
+    ///
+    /// ```text
+    ///    .             .
+    ///   / \           / \
+    /// p3   .        v3   .
+    ///     / \           / \
+    ///   p2   .   ↔    v2   .
+    ///       / \           / \
+    ///     p1   _        v1  ()
+    ///
+    /// p3 ↔ v3
+    /// p2 ↔ v2
+    /// p1 ↔ v1
+    /// ```
+    variables: Vec<Pattern>,
+    /// For each scope, the number of patterns from the outer scope.
+    ///
+    /// When a scope is left, the pattern stack is truncated to this number.
+    scopes: Vec<usize>,
```

Comment:
I think I want to 257ab7f5311ab39bcf4de24d9416baa525c55db7. The `scopes` member basically turns the `variables` member back into `Vec<Vec<Pattern>>`. The remaining changes of the PR stay intact.